keps/sig-apps/2307-job-tracking-without-lingering-pods/README.md

```diff
@@ -307,21 +319,95 @@ finalizer.
 The job controller adds the finalizer in the same patch request that modifies
 the owner reference.
 
+## Monitoring Pods with finalizers
+
+Starting in 1.26, the metric `job_terminated_pod_tracking_finalizer` is a gauge
+that tracks the number of terminated pods (`.status.phase=(Succeeded|Failed)`)
+that currently have a job tracking finalizer.
+
+The job controller tracks this metric in its event handlers.
+
+## Migrating Jobs with legacy tracking
+
+Once `JobTrackingWithFinalizers` graduates to stable, Jobs that start in a
+kubernetes version where `JobTrackingWithFinalizer` is disabled need to be
+migrated to the new tracking. This migration mechanism will be initially guarded
+by the feature gate `MigrateJobLegacyTracking`, starting in 1.26,
+enabled by default.
+
+When the feature gate `MigrateJobLegacyTracking` is enabled, the job controller
+migrates jobs with legacy tracking to tracking with finalizers as described
+below:
+
+If a Job doesn't have the annotation `batch.kubernetes.io/job-completion`, it
+means that the Job is not currently tracked with finalizers. The job controller
+starts the following migration process:
+1. Add the finalizer `batch.kubernetes.io/job-completion` to all pods with
+   `.status.phase=(Pending/Running)`.
+2. Ignore pods with `.status.phase=(Complete/Failed)` that don't have the `batch.kubernetes.io/job-completion`.
+   They are considered to be already counted in `.status.(failed/succeeded)`.
+3. Add the annotation `batch.kubernetes.io/job-completion`.
```

Comment:
This bit is probably the only one which I have some doubts around. Can you estimate how many jobs might be using the old mechanism assuming both cases:
1. with the feature on by default in 1.25.
2. with the feature off in 1.25. 

From my quick calculations, the 2nd case will be the most impactful, since that's basically all the jobs using the old mechanism that need to be migrated. Is there something we can do here to make the pain to users to be smaller? Maybe advise them to run 1.25 with the feature on before the upgrade? Would that kind of approach make it more reliable?

Reply:
Should we allow users to disable `MigrateJobLegacyTracking` at all?

Reply:
Or maybe we should do it gradually, iow. new jobs should use only the new tracking, but leave the old ones as they where?

Reply:
In general, I have no way to know.

For case 1, and assuming that admins are not doing "binge upgrading", it should be close to zero. There are 4 months between releases, which should be plenty of time for even the slowest jobs to finish.

Case 2 can be directly impacted.

I'm thinking that the recommendation should be either of these:
1. If JobTrackingWithFinalizers was enabled in 1.25, enable MigrateJobLegacyTracking in 1.26.
2. If the feature was disabled in 1.24, disable MigrateJobLegacyTracking in 1.26. They remaining jobs will be migrated in 1.27.

We could also never introduce the code for migration and document this:
If a Job that was created in 1.25- with JobTrackingWithFinalizers disabled, and it is still running in 1.26, the job controller would recreate all its pods upon upgrade to 1.27.

For people that had JobTrackingWithFinalizers enabled in 1.25, that gives them 2 releases of buffer for their jobs to finish.

Reply:
I think the migration code is an overkill, the code is too complex and this will only increase sources of bugs/errors. I think it is sufficient to have a release note indicating that upgrading clusters back to back through two releases may disrupt running jobs.

Reply:
Removed

Reply:
Totally agree with Abdullah about that migration being overkill (and potential source of unwanted and unexpected problems). All we should do is to ensure good communication about these upcoming changes.